From 6342842b69302ad5530b07d8390102b0648f4b9a Mon Sep 17 00:00:00 2001
From: Franz Schnyder <franz.schnyder@toradex.com>
Date: Thu, 21 May 2026 11:29:47 +0200
Subject: [PATCH] mlinux: moal_main: fix device wakeup capability

Currently, the device is always advertised as being capable of
waking up, even if no interrupt line has been allocated for it.

Initialize the device only as wakeup capable if an interrupt has been
registered.

Upstream-Status: Pending
Signed-off-by: Franz Schnyder <franz.schnyder@toradex.com>
---
 mlinux/moal_main.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/mlinux/moal_main.c b/mlinux/moal_main.c
index 90a7342..2f49e06 100644
--- a/mlinux/moal_main.c
+++ b/mlinux/moal_main.c
@@ -14819,6 +14819,11 @@ void woal_regist_oob_wakeup_irq(moal_handle *handle)
 		goto err_exit;
 	}
 
+	if (device_init_wakeup(dev, true)) {
+		dev_err(dev, "Failed to init wakeup for mwifiex\n");
+		goto err_exit;
+	}
+
 	disable_irq(handle->irq_oob_wakeup);
 
 	LEAVE();
@@ -14840,8 +14845,10 @@ void woal_unregist_oob_wakeup_irq(moal_handle *handle)
 	struct device *dev = handle->hotplug_device;
 
 	ENTER();
-	if (handle->irq_oob_wakeup >= 0)
+	if (handle->irq_oob_wakeup >= 0) {
+		device_init_wakeup(dev, false);
 		devm_free_irq(dev, handle->irq_oob_wakeup, handle);
+	}
 	LEAVE();
 }
 
@@ -15457,7 +15464,6 @@ moal_handle *woal_add_card(void *card, struct device *dev, moal_if_ops *if_ops,
 		PRINTM(MFATAL, "Firmware Init Failed\n");
 		goto err_init_fw;
 	}
-	device_init_wakeup(dev, true);
 #ifdef SD8887
 	if (IS_SD8887(handle->card_type)) {
 		if (handle->fw_release_number.minorRevNum == 75) {
-- 
2.43.0

